### PR TITLE
Issue577 Implement ERC20 enterprise tests

### LINKED ISSUE #577: Implement ERC20 Enterprise Template for V4
Implement ERC20 Enterprise Template for V4 from [here](https://github.com/oceanprotocol/contracts/blob/v4main/test/unit/datatokens/ERC20TemplateEnterprise.test.js).

### PR DESCRIPTION
**Pull request by trizin**
Fixes #577  .

Changes proposed in this PR:

- Implemented ERC20 enterprise tests.
- Added `create_dispenser` function to ERC20Enterprise class.

### CODE REVIEW

**Review by mariacarmina — COMMENTED** [review]
Please resolve the conflicts

File: ocean_lib/models/v4/test/test_erc20_enterprise.py [hunk 1]
```diff
@@ -235,7 +235,7 @@ def test_buy_from_fre_and_order(
     )
     mock_dai_contract.approve(
         spender=erc20_enterprise_token.address,
-        amount=consume_fee_amount * 2,
+        amount=web3.toWei(1000, "ether"),
```
Comment (mariacarmina) [inline ocean_lib/models/v4/test/test_erc20_enterprise.py:238]: why did you modify this? we can set either to 2**256 -1, either to the exact amount.

File: ocean_lib/models/v4/erc20_enterprise.py [hunk 2]
```diff
@@ -25,3 +26,20 @@ def buy_from_dispenser_and_order(
         return self.send_transaction(
             "buyFromDispenserAndOrder", (order_params, dispenser_address), from_wallet
         )
+
+    def create_dispenser(
```
Comment (mariacarmina) [inline ocean_lib/models/v4/erc20_enterprise.py:30]: please delete this function, it is inherited from ERC20.

File: ocean_lib/models/v4/test/test_erc20_enterprise.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    mock_usdc_contract = ERC20Token(web3, get_address_of_type(config, "MockUSDC"))
+    mock_dai_contract = ERC20Token(web3, get_address_of_type(config, "MockDAI"))
+    fixed_rate_exchange = FixedRateExchangeV4(
+        web3, get_address_of_type(config, "FixedPrice")
+    )
+
+    _, erc20_enterprise_token = deploy_erc721_erc20(
+        web3=web3,
+        config=config,
+        cap=web3.toWei(100, "ether"),
```
Comment (mariacarmina) [inline ocean_lib/models/v4/test/test_erc20_enterprise.py:159]: please reorder the parameters.

File: ocean_lib/models/v4/test/test_erc20_enterprise.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    assert tx_receipt["status"] == 1
+
+    new_fixed_rate_event = erc20_enterprise_token.get_event_log(
+        "NewFixedRate",
+        from_block=tx_receipt.blockNumber,
+        to_block=web3.eth.block_number,
+        filters=None,
+    )
+
+    exchange_id = new_fixed_rate_event[0]["args"]["exchangeId"]
```
Comment (mariacarmina) [inline ocean_lib/models/v4/test/test_erc20_enterprise.py:192]: you can access the `exchange_id` with `new_fixed_rate_event[0].args.exchangeId`.

File: ocean_lib/models/v4/test/test_erc20_enterprise.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+        uints=[18, 18, web3.toWei(1, "ether"), web3.toWei(1, "ether"), 1],
+    )
+
+    tx = erc20_enterprise_token.create_fixed_rate(
+        fixed_data=fixed_data,
+        from_wallet=consumer_wallet,
+    )
+
+    tx_receipt = web3.eth.wait_for_transaction_receipt(tx)
+    assert tx_receipt["status"] == 1
```
Comment (mariacarmina) [inline ocean_lib/models/v4/test/test_erc20_enterprise.py:183]: tx_receipt is an AttributeDict, you can access the status like that `tx_receipt.status`.

File: ocean_lib/models/v4/test/test_erc20_enterprise.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+    balance_opf_publish_before = mock_usdc_contract.balanceOf(opf_collector_address)
+
+    tx = erc20_enterprise_token.buy_from_fre_and_order(
+        order_params=order_params,
+        fre_params=fre_params,
+        from_wallet=consumer_wallet,
+    )
+
+    tx_receipt = web3.eth.wait_for_transaction_receipt(tx)
+    assert tx_receipt["status"] == 1
```
Comment (mariacarmina) [inline ocean_lib/models/v4/test/test_erc20_enterprise.py:271]: you can use `tx_receipt.status`.

File: ocean_lib/models/v4/test/test_erc20_enterprise.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    web3, config, consumer_wallet, publisher_wallet, factory_deployer_wallet
+):
+    mock_usdc_contract = ERC20Token(web3, get_address_of_type(config, "MockUSDC"))
+    mock_dai_contract = ERC20Token(web3, get_address_of_type(config, "MockDAI"))
+    dispenser = DispenserV4(web3, get_address_of_type(config, "Dispenser"))
+
+    _, erc20_enterprise_token = deploy_erc721_erc20(
+        web3=web3,
+        config=config,
+        cap=web3.toWei(100, "ether"),
```
Comment (mariacarmina) [inline ocean_lib/models/v4/test/test_erc20_enterprise.py:27]: can you reorder the params here as well for consistency purpose?

File: ocean_lib/models/v4/test/test_erc20_enterprise.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+    dispenser_data = DispenserData(
+        dispenser_address=dispenser.address,
+        allowed_swapper=ZERO_ADDRESS,
+        max_balance=web3.toWei(1, "ether"),
+        max_tokens=web3.toWei(1, "ether"),
+    )
+    tx = erc20_enterprise_token.create_dispenser(
+        dispenser_data=dispenser_data, from_wallet=consumer_wallet, with_mint=True
+    )
+    tx_receipt = web3.eth.waitForTransactionReceipt(tx)
```
Comment (mariacarmina) [inline ocean_lib/models/v4/test/test_erc20_enterprise.py:43]: change this web3 deprecated function with `wait_for_transaction_receipt`.

**Review by mariacarmina — CHANGES_REQUESTED** [review]

File: ocean_lib/models/v4/erc20_enterprise.py (hunk 2 above)
Comment (trizin) [reply ocean_lib/models/v4/erc20_enterprise.py:30]: Unfortunately, I cannot use the `create_dispenser` function from the ERC20. The parameters for the `create_dispenser` function in enterprise ERC20 are different than the ERC20.

File: ocean_lib/models/v4/test/test_erc20_enterprise.py [hunk 9]
```diff
@@ -38,23 +39,23 @@ def test_buy_from_dispenser_and_order(
         max_tokens=web3.toWei(1, "ether"),
     )
     tx = erc20_enterprise_token.create_dispenser(
-        dispenser_data=dispenser_data, from_wallet=consumer_wallet, with_mint=True
+        dispenser_data=dispenser_data, from_wallet=publisher_wallet, with_mint=True
```
Comment (mariacarmina) [inline ocean_lib/models/v4/test/test_erc20_enterprise.py:42]: Can we reorder here again the parameters?

**Review by mariacarmina — COMMENTED** [review]
Just a final tweak and in rest it looks great!

**Review by mariacarmina — APPROVED** [review]

**Pull request merged**